In [104]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [105]:
titanic_data = pd.read_csv('data/titanic_original.csv')

In [106]:
titanic_data

pclass  survived                                             name  \
0        1.0       1.0                    Allen, Miss. Elisabeth Walton   
1        1.0       1.0                   Allison, Master. Hudson Trevor   
2        1.0       0.0                     Allison, Miss. Helen Loraine   
3        1.0       0.0             Allison, Mr. Hudson Joshua Creighton   
4        1.0       0.0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1305     3.0       0.0                            Zabour, Miss. Thamine   
1306     3.0       0.0                        Zakarian, Mr. Mapriededer   
1307     3.0       0.0                              Zakarian, Mr. Ortin   
1308     3.0       0.0                               Zimmerman, Mr. Leo   
1309     NaN       NaN                                              NaN   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000    0.0    0.0   24160  211.3375       B5        S    2   
1       male   0.9167    1.0    2.0  113781  151.5500  C22 C26        S   11   
2     female   2.0000    1.0    2.0  113781  151.5500  C22 C26        S  NaN   
3       male  30.0000    1.0    2.0  113781  151.5500  C22 C26        S  NaN   
4     female  25.0000    1.0    2.0  113781  151.5500  C22 C26        S  NaN   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1305  female      NaN    1.0    0.0    2665   14.4542      NaN        C  NaN   
1306    male  26.5000    0.0    0.0    2656    7.2250      NaN        C  NaN   
1307    male  27.0000    0.0    0.0    2670    7.2250      NaN        C  NaN   
1308    male  29.0000    0.0    0.0  315082    7.8750      NaN        S  NaN   
1309     NaN      NaN    NaN    NaN     NaN       NaN      NaN      NaN  NaN   

       body                        home.dest  
0       NaN                     St Louis, MO  
1       NaN  Montreal, PQ / Chesterville, ON  
2       NaN  Montreal, PQ / Chesterville, ON  
3     135.0  Montreal, PQ / Chesterville, ON  
4       NaN  Montreal, PQ / Chesterville, ON  
...     ...                              ...  
1305    NaN                              NaN  
1306  304.0                              NaN  
1307    NaN                              NaN  
1308    NaN                              NaN  
1309    NaN                              NaN  

[1310 rows x 14 columns]

In [107]:
titanic_data.describe()

pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881135     0.498854     0.385027   
std       0.837836     0.486055    14.413500     1.041658     0.865560   
min       1.000000     0.000000     0.166700     0.000000     0.000000   
25%       2.000000     0.000000    21.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    39.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare        body  
count  1308.000000  121.000000  
mean     33.295479  160.809917  
std      51.758668   97.696922  
min       0.000000    1.000000  
25%       7.895800   72.000000  
50%      14.454200  155.000000  
75%      31.275000  256.000000  
max     512.329200  328.000000

In [108]:
titanic_data.body.describe()

count    121.000000
mean     160.809917
std       97.696922
min        1.000000
25%       72.000000
50%      155.000000
75%      256.000000
max      328.000000
Name: body, dtype: float64

The "Titanic" dataset (located at https://raw.githubusercontent.com/Geoyi/Cleaning-Titanic-Data/master/titanic_original.csv) contains information about 887 passengers on the Titanic (out of 2224 total). Which column has the most missing values?

- body

In [109]:
iris_data = pd.read_csv('data/iris.data.csv')

In [110]:
iris_data

sepal_length  sepal_width  petal_length  petal_width      iris_class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]

Observations and Features, Part 1
The Iris dataset is one of the most populat datasets in data processing. You can find the data and description at https://archive.ics.uci.edu/ml/datasets/iris. How many observations are there?

- 150

Observations and Features, Part 2
And how many features?

- 5

In [111]:
iris_data.iris_class.unique()

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

Target Values
The last column contains the different species of iris plant. How many species are there?

 - 3

In [26]:
iris_data.sepal_length.dtype

dtype('float64')

In [54]:
filtered_data = iris_data[(iris_data.sepal_length >= 4.5) & (iris_data.sepal_length <= 10)]


In [59]:
filtered_data.sort_values(['sepal_length'])

sepal_length  sepal_width  petal_length  petal_width      iris_class
41            4.5          2.3           1.3          0.3     Iris-setosa
22            4.6          3.6           1.0          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
47            4.6          3.2           1.4          0.2     Iris-setosa
6             4.6          3.4           1.4          0.3     Iris-setosa
..            ...          ...           ...          ...             ...
118           7.7          2.6           6.9          2.3  Iris-virginica
135           7.7          3.0           6.1          2.3  Iris-virginica
122           7.7          2.8           6.7          2.0  Iris-virginica
117           7.7          3.8           6.7          2.2  Iris-virginica
131           7.9          3.8           6.4          2.0  Iris-virginica

[146 rows x 5 columns]

In [60]:
filtered_data.petal_length.mean()

3.826712328767123

Iris Data Insights
Take all Iris flowers whose sepal length is between 4.5 and 10 (inclusive). What is the average petal length of these flowers (in cm)? Round your answer to two decimal places. Write your answer with a decimal point, for example "4.5".

- 3.83

In [61]:
red_wine_data = pd.read_csv('data/winequality-red.csv', sep=';')

In [62]:
red_wine_data

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6  

[1599 rows x 12 columns]

In [63]:
red_wine_data.quality.unique()

array([5, 6, 7, 4, 8, 3])

In [64]:
red_wine_data.quality.dtype

dtype('int64')

In [65]:
red_wine_data.quality.mean()

5.6360225140712945

In [73]:
white_wine_data = pd.read_csv('data/winequality-white.csv', sep=';')

In [74]:
white_wine_data

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0              0.27         0.36            20.7      0.045   
1               6.3              0.30         0.34             1.6      0.049   
2               8.1              0.28         0.40             6.9      0.050   
3               7.2              0.23         0.32             8.5      0.058   
4               7.2              0.23         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
0         8.8        6  
1         9.5        6  
2        10.1        6  
3         9.9        6  
4         9.9        6  
...       ...      ...  
4893     11.2        6  
4894      9.6        5  
4895      9.4        6  
4896     12.8        7  
4897     11.8        6  

[4898 rows x 12 columns]

In [75]:
white_wine_data.quality.unique()

array([6, 5, 7, 8, 4, 3, 9])

In [76]:
white_wine_data.quality.dtype

dtype('int64')

In [78]:
white_wine_data.quality.mean()

5.87790935075541

In [80]:
if red_wine_data.quality.mean() > white_wine_data.quality.mean():
    print('Red wine has higher quality')
elif red_wine_data.quality.mean() < white_wine_data.quality.mean():
    print('White wine has higher quality')
else:
    print('Both wines has equal quality')

White wine has higher quality


The dataset here: http://archive.ics.uci.edu/ml/datasets/Wine+Quality contains information about different red and white wines. Read the two files into Python. Which kinds of wine seem to be preferred (i.e. have greater average quality)? Write "red" or "white" in the textbox.
- white

In [88]:
red_wine_data.quality.corr(red_wine_data.alcohol)

0.4761663240011361

In [89]:
red_wine_data.alcohol.corr(red_wine_data.quality)

0.4761663240011362

Quality and Alcohol Content, Part 1
Does higher alcohol content mean better quality? We can't say directly, but we might explore the data. What is the correlation of alcohol content to quality for red wines? Round your answer to two decimal places.
- 0.48

In [90]:
white_wine_data.quality.corr(white_wine_data.alcohol)


0.435574715461373

In [91]:
white_wine_data.alcohol.corr(white_wine_data.quality)

0.43557471546137305

Quality and Alcohol Content, Part 2
What is the correlation of alcohol content to quality for white wines? Round your answer to two decimal places.
- 0.44